### PR TITLE
feat: add c implementation of `stats/base/dists/rayleigh/logpdf`

### LINKED ISSUE #3802: [RFC]: Add C implementation for `@stdlib/stats/base/dists/rayleigh/logpdf`
### Description

This RFC proposes adding a C implementation, including a Node.js native addon, C benchmarks, and C examples, for the [`@stdlib/stats/base/dists/rayleigh/logpdf`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf) package.

When adding support, the following tasks should be completed:

1. A C source implementation should be added to a `src` folder within the package's directory.
2. A native add-on interface should be added in order to call the C implementation from JavaScript.
3. A JavaScript file should be added to the `lib` folder to allow benchmarking and unit testing the native implementation from JavaScript.
4. A new test file should be created which includes unit tests testing against expected behavior.
5. C benchmarks should be added which measure performance for the C source implementation.
6. A file containing C examples should be added to demonstrate example usage.
7. JavaScript benchmarks should be added which measure performance for the native add-on.
8. Build and configuration files should be added to allow compiling and running the C source implementation using project `make` commands (as documented below).
9. The README should be updated to include documentation for the added C API.

To provide a concrete example of what a PR adding the desired method should contain, see https://github.com/stdlib-js/stdlib/pull/3354, which is a PR adding a C implementation for the CDF of an arcsine distribution. This should provide an idea of what is expected.

#### Prerequisites

1. Review JavaScript implementation in `lib/main.js` and its required modules to identify all `stdlib` functions used.
2. Verify C implementations exist for each required function.

### Related Issues

#3355

### Questions

No.

### Other

Once the implementation is ready including C examples and benchmarks, we should be able to run the following `make` commands.

#### Build native add-on

```bash
NODE_ADDONS_PATTERN="@stdlib/stats/base/dists/rayleigh/logpdf" make install-node-addons
```

#### Run C examples

```bash
make examples-c-files FILES="$(pwd)/lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/examples/c/example.c"
```

#### Run C benchmarks

```bash
make benchmark-c-files FILES="$(pwd)/lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/benchmark/c/benchmark.c"
```

#### Run JavaScript benchmarks

```bash
make benchmark-javascript-files FILES="$(pwd)/lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/benchmark/benchmark.native.js"
```

#### Run all tests

```bash
make TESTS_FILTER=".*/@stdlib/stats/base/dists/rayleigh/logpdf/.*" test
```

### Notes

- If you are interested in contributing a PR which addresses this RFC and are a **first-time contributor** or not yet familiar with our project conventions, please minimize the submission of LLM-generated code. Please consult our [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md) and the associated [development guide](https://github.com/stdlib-js/stdlib/blob/develop/docs/development.md). Failure to respect project conventions will result in your PR being rejected without review. Thank you for understanding!

### Checklist

- [x] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [x] Searched for existing issues and pull requests.
- [x] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by Vinit-Pandit**
Resolves #3802 

## Description

> What is the purpose of this pull request?

This pull request:

-   add c implementation of `stats/base/dists/rayleigh/logpdf`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #3802

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/benchmark/benchmark.js [hunk 1]
```diff
@@ -32,15 +33,21 @@ var logpdf = require( './../lib' );
...
+	sigma = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		x[ i ] = ( randu() * 100.0 ) - 100.0;
+		sigma[ i ] = ( randu() * 20.0 ) + EPS;
+	}
 	for ( i = 0; i < b.iterations; i++ ) {
-		x = ( randu()*100.0 ) - 100;
-		sigma = ( randu()*20.0 ) + EPS;
-		y = logpdf( x, sigma );
+		y = logpdf( x[ i % 100 ], sigma[ i % 100 ] );
```
Comment (Vinit-Pandit) [inline lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/benchmark/benchmark.js]:
```suggestion
		y = logpdf( x[ i % len ], sigma[ i % len ] );
```

File: lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/benchmark/benchmark.native.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	len = 100;
+	b.tic();
+	x = new Float64Array( len );
+	sigma = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		x[ i ] = ( randu() * 100.0 ) - 100.0;
+		sigma[ i ] = ( randu() * 20.0 ) + EPS;
+	}
+	for ( i = 0; i < b.iterations; i++ ) {
+		y = logpdf( x[ i % 100 ], sigma[ i % 100 ] );
```
Comment (Vinit-Pandit) [inline lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/benchmark/benchmark.native.js]:
```suggestion
		y = logpdf( x[ i % len ], sigma[ i % len ] );
```

File: lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/lib/native.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+// MODULES //
+
+var addon = require( './../src/addon.node' );
+
+
+// MAIN //
+
+/**
+* Evaluates the logarithm of the probability density function (PDF) for a Rayleigh distribution with scale parameter `sigma` at a value `x`.
+*
```
Comment (Vinit-Pandit) [inline lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/lib/native.js:30]:
```suggestion
*
* @private
```

File: lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/benchmark/c/benchmark.c [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+static double benchmark( void ) {
+	double elapsed;
+	double x[ 100 ];
+	double sigma[ 100 ];
+	double y;
+	double t;
+	int i;
+
+	for ( i = 0; i < 100; i++ ) {
+		x[ i ] = random_uniform( 20, 20 + STDLIB_CONSTANT_FLOAT64_EPS );
```
Comment (Vinit-Pandit) [inline lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/benchmark/c/benchmark.c]:
```suggestion
		x[ i ] = random_uniform( STDLIB_CONSTANT_FLOAT64_EPS, STDLIB_CONSTANT_FLOAT64_EPS  + 20.0 );
```

File: lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/src/main.c [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+* limitations under the License.
+*/
+
+#include "stdlib/stats/base/dists/rayleigh/logpdf.h"
+#include "stdlib/math/base/assert/is_nan.h"
+#include "stdlib/math/base/special/pow.h"
+#include "stdlib/math/base/special/ln.h"
+#include "stdlib/constants/float64/pinf.h"
+#include "stdlib/constants/float64/ninf.h"
+
```
Comment (Vinit-Pandit) [inline lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/src/main.c:25]:
```suggestion
```

File: lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/benchmark/benchmark.js [hunk 6]
```diff
@@ -56,17 +63,22 @@ bench( pkg, function benchmark( b ) {
...
 	var y;
 	var i;
 
+	len = 100;
 	sigma = 4.0;
+	x = new Float64Array( len );
 	mylogpdf = logpdf.factory( sigma );
+	for ( i = 0; i < len; i++ ) {
+		x[ i ] = ( randu() * 50.0 ) - 25.0;
+	}
```
Comment (Vinit-Pandit) [inline lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/benchmark/benchmark.js]:
```suggestion
bench( pkg+':factory', function benchmark( b ) {
	var mylogpdf;
	var sigma;
	var x;
	var y;
	var i;

	sigma = 4.0;
	mylogpdf = logpdf.factory( sigma );
```

File: lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/benchmark/benchmark.js [hunk 7]
```diff
@@ -56,17 +63,22 @@ bench( pkg, function benchmark( b ) {
...
 
 	b.tic();
 	for ( i = 0; i < b.iterations; i++ ) {
-		x = ( randu()*50.0 ) - 25.0;
-		y = mylogpdf( x );
+		y = mylogpdf( x[ i % len ] );
```
Comment (Vinit-Pandit) [inline lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/benchmark/benchmark.js]:
```suggestion
		x = ( randu()*50.0 ) - 25.0;
		y = mylogpdf( x );
```

File: lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/benchmark/c/benchmark.c [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+#include "stdlib/stats/base/dists/rayleigh/logpdf.h"
+#include "stdlib/constants/float64/eps.h"
+#include <stdlib.h>
+#include <stdio.h>
+#include <math.h>
+#include <time.h>
+#include <sys/time.h>
+
+#define NAME "rayleigh-cdf"
```
Comment (Vinit-Pandit) [inline lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/benchmark/c/benchmark.c]:
```suggestion
#define NAME "rayleigh-logpdf"
```

**Comment by Vinit-Pandit** [comment]
Wops ! really sorry for this [blunder](https://github.com/stdlib-js/stdlib/pull/3966/commits/50f321c7bb32c1aaf3db1bf097bcbb7b864c4535)

File: lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/README.md [hunk 9]
```diff
@@ -153,6 +153,101 @@ for ( i = 0; i < 10; i++ ) {
...
+    double sigma;
+    double x;
+    double y;
+    int i;
+
+    for ( i = 0; i < 25; i++ ) {
+        x = random_uniform( 0.0, 10.0 );
+        sigma = random_uniform( 0.0, 10.0 );
+        y = stdlib_base_dists_rayleigh_logpdf( x, sigma );
+        printf( "x: %lf, σ: %lf, f(x;σ): %lf\n", x, sigma, y );
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/README.md]:
```suggestion
        printf( "x: %lf, σ: %lf, ln(f(x;σ)): %lf\n", x, sigma, y );
```

File: lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/examples/c/example.c [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+	double sigma;
+	double x;
+	double y;
+	int i;
+
+	for ( i = 0; i < 25; i++ ) {
+		x = random_uniform( 0.0, 10.0 );
+		sigma = random_uniform( 0.0, 10.0 );
+		y = stdlib_base_dists_rayleigh_logpdf( x, sigma );
+		printf( "x: %lf, σ: %lf, f(x;σ): %lf\n", x, sigma, y );
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/rayleigh/logpdf/examples/c/example.c]:
```suggestion
		printf( "x: %lf, σ: %lf, ln(f(x;σ)): %lf\n", x, sigma, y );
```

**Review by Planeshifter — APPROVED** [review]
Thank you! PR is looking good and should land shortly.

**Pull request merged**